### DISCUSSION TITLE
with materialized cte

### DISCUSSION THREAD

**Discussion by k3nj1g** [jkk/honeysql#391] [Q&A] [answered]
Can you help with register clause for materialized cte? For query like:

```
WITH foo AS MATERIALIZED (
  SELECT foo 
  FROM bar
)
```

**Comment by seancorfield** [comment, marked as answer]
This is a tricky one because `MATERIALIZED` here is a hint on the CTE clause rather than a specific clause itself -- and CTE clauses are always wrapped in `(`..`)` so you would pretty much need a full `:with-materialized` clause that duplicated most of what's already in the `:with` formatter with just a slight change.

It would probably be easier all around if I just added support for this in core.

This might be a good opportunity to add support for metadata directly into the formatter though, since the natural way to deal with SQL hints would be through metadata on the clauses I think:

```clojure
dev=> (sql/format {:with [[:foo ^:materialized {:select :foo :from :bar}]]})
["WITH foo AS MATERIALIZED (SELECT foo FROM bar)"]
```

(although, in this case, that means the metadata support has to be in the `:with` formatter, not the general clause formatter, so I'll need to give this some thought)
